### PR TITLE
[GSK-1426] Use WebSocket to replace GRPC for worker-backend communication

### PR DESCRIPTION
**Pull request by Inokinoki**
## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [x] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [ ] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CODE_OF_CONDUCT.md) document.
- [ ] I've read the [`CONTRIBUTING.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CONTRIBUTING.md) guide.
- [ ] I've updated the code style using `make codestyle`.
- [ ] I've written tests for all new methods and classes that I created.
- [ ] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: backend/src/main/java/ai/giskard/service/ModelService.java [hunk 1]
```diff
@@ -37,74 +42,121 @@ public class ModelService {
...
+            );
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        } catch (NullPointerException e) {
+            throw new NullPointerException("Did not get an validate ML Worker RunModelForDataFrame reply");
+        }
+        if (result instanceof MLWorkerWSRunModelForDataFrameDTO response) {
+            return response;
+        }
+        throw new NullPointerException("Cannot get ML Worker RunModelForDataFrame reply");
```
Comment (andreybavt) [inline backend/src/main/java/ai/giskard/service/ModelService.java:104]: I don't get this NPE usage (and also in other places where you added it). There's no null pointer in this context, you're just reacting to an unexpected result. You should create a subclass of `RuntimeException` and throw it if needed.

File: backend/src/main/java/ai/giskard/web/rest/controllers/testing/TestController.java [hunk 2]
```diff
@@ -44,27 +50,40 @@ public TestTemplateExecutionResultDTO runAdHocTest(@RequestBody RunAdhocTestRequ
...
-                    .buildTestArgument(arguments, input.getName(), input.getValue(), project.getKey(),
-                        giskardMapper.fromDTO(input).getParams(), sample));
+            MLWorkerWSBaseDTO result = null;
+            try {
+                result = mlWorkerWSCommService.performAction(
+                    workerID,
+                    MLWorkerWSAction.runAdHocTest,
+                    param
+                );
+            } catch (JsonProcessingException e) {
```
Comment (andreybavt) [inline backend/src/main/java/ai/giskard/web/rest/controllers/testing/TestController.java:71]: we should handle these errors on the `performAction` level

**Review by andreybavt — CHANGES_REQUESTED** [review]

**Comment by Inokinoki** [comment]
How about separating the DTOs, the tests from the communication part, in some other PRs?

**Comment by Inokinoki** [comment]
> How about separating the DTOs, the tests from the communication part, in some other PRs?

Rebased to remove commits for unit tests. Commits picked at https://github.com/Giskard-AI/giskard/pull/1277

**Comment by Inokinoki** [comment]
HF access token commits are separated for easier review: https://github.com/Giskard-AI/giskard/pull/1278

File: python-client/giskard/ml_worker/websocket/listener.py [hunk 3]
```diff
@@ -0,0 +1,838 @@
...
+            cellLevel=test.cell_level,
+            columnType=test.column_type,
+            processType=test.process_type.name,
+        )
+        for test in tests_registry.get_all().values()
+        if test.type == callable_type
+    }
+
+
+@websocket_actor(MLWorkerAction.getInfo)
```
Comment (andreybavt) [inline python-client/giskard/ml_worker/websocket/listener.py:319]: this file is a bit huge, I'd split it by low level utilities (actiond_in_tread, etc.) and the actors themselves. Plus you can group actors into separate submodules to keep it more organized

File: backend/src/main/java/ai/giskard/config/WebSocketSecurityConfig.java [hunk 4]
```diff
@@ -63,6 +90,28 @@ public Message<?> preSend(Message<?> message, MessageChannel channel) {
...
+                if (StompCommand.SUBSCRIBE.equals(accessor.getCommand())) {
+                    String destination = accessor.getDestination();
+                    if (destination == null)
+                        throw new AccessDeniedException("Cannot find available worker");
+                    if (!destination.startsWith("/ml-worker/"))
+                        return message; // By-pass the other subscriptions
+                    if (destination.equals(WebSocketConfig.INTERNAL_ML_WORKER_TOPIC) &&
+                        mlWorkerWSService.associateWorker(MLWorkerID.INTERNAL.toString(),
+                            Objects.requireNonNull(accessor.getSessionId())) ||
+                        destination.equals(WebSocketConfig.EXTERNAL_ML_WORKER_TOPIC) &&
```
[the remark this answers is not in the document]
Comment (Inokinoki) [reply backend/src/main/java/ai/giskard/config/WebSocketSecurityConfig.java:103]: Same to me :) This was just to make SonarLint happy

File: python-client/giskard/ml_worker/websocket/listener.py (hunk 3 above)
Comment (Inokinoki) [reply python-client/giskard/ml_worker/websocket/listener.py:180]: Done, but actors can be organised in further steps.

File: backend/src/main/java/ai/giskard/config/WebSocketSecurityConfig.java [hunk 5]
```diff
@@ -47,6 +56,24 @@ public Message<?> preSend(Message<?> message, MessageChannel channel) {
...
                 }
                 if (StompCommand.CONNECT.equals(accessor.getCommand())) {
+                    log.debug("Session {} CONNECT", accessor.getSessionId());
+                    // Check if an internal worker trying to connect
+                    List<String> internalTokenHeaders = accessor.getNativeHeader("token");
+                    if (internalTokenHeaders != null && !internalTokenHeaders.isEmpty()
+                        && StringUtils.hasText(internalTokenHeaders.get(0))) {
+                        // Validate the token
+                        try {
+                            String token = Files.readString(Paths.get(fileLocationService.giskardHome().toString(), "run", "internal-ml-worker"));
```
[the remark this answers is not in the document]
Comment (Inokinoki) [reply backend/src/main/java/ai/giskard/config/WebSocketSecurityConfig.java:66]: Or can I say it is a pending internal worker, until the worker has subscribed the topic?

**Comment by andreybavt** [comment]
@Inokinoki can you also fix sonar complaints please? Those that are not relevant can be ignored (by adding NOSONAR for example)

File: backend/src/main/java/ai/giskard/config/WebSocketSecurityConfig.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreybavt) [reply backend/src/main/java/ai/giskard/config/WebSocketSecurityConfig.java:66]: I'd just simply and call it simply "internal worker". According to our communication flow "SUBSCRIBE" comes right after "CONNECT", doesn't it? In this case it should be an exceptional case if there's a CONNECT but no SUBSCRIBE

File: backend/src/main/java/ai/giskard/service/ml/MLWorkerWSCommService.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        HashMap<String, Object> data = new HashMap<>();
+        data.put("config", "MAX_STOMP_ML_WORKER_REPLY_SIZE");
+        data.put("value", maxPayloadLength);
+        simpMessagingTemplate.convertAndSend(
+            String.join("/", WebSocketConfig.ML_WORKER_TOPIC_PREFIX, workerID.toString(), WebSocketConfig.ML_WORKER_CONFIG_TOPIC),
+            data
+        );
+    }
+
+    @Transactional(propagation = Propagation.NEVER)
```
[the remark this answers is not in the document]
Comment (andreybavt) [reply backend/src/main/java/ai/giskard/service/ml/MLWorkerWSCommService.java:44]: @Inokinoki

**Pull request merged**